### PR TITLE
Refactor: pkg/util into sub-pkgs

### LINKED ISSUE #15634: pkg/util should contain no code
we should move everything into sub-pkgs.

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
- move code in pkg/util into sub-pkgs
- delete some unused funcs

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #15634

**Special notes for your reviewer**:
This is the final work of #15634. It will close that issue.
/cc @thockin 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by deads2k** [comment]
@k8s-bot ok to tests

I agree with the goal.  I think we should reassess the general utility of some of these, but this is a step in the right direction.  

@xiangpengzhao please squash.

**Comment by xiangpengzhao** [comment]
>@k8s-bot ok to tests

Tests aren't triggered. Bot wants `ok to test` :)

>I agree with the goal. I think we should reassess the general utility of some of these, but this is a step in the right direction.

We also need to remove the empty pkgs such as util/intstr when they're no longer needed.

>please squash.

I split this PR into two commits to make it easy to review. I will squash them when reviewers think the PR lgtm.

**Comment by thockin** [comment]
/approve

On principle I am approving, but will assign for code review.

File: pkg/util/pointer/pointer.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Comment (mengqiy) [inline pkg/util/pointer/pointer.go]: s/2014/2017
Same for the all the new files in the new packages.

**Comment by xiangpengzhao** [comment]
Rebased and inline comment addressed. PTAL. @mengqiy

**Comment by mengqiy** [comment]
LGTM
Please squash.

**Comment by xiangpengzhao** [comment]
squashed. PTAL. @mengqiy

**Comment by mengqiy** [comment]
@xiangpengzhao Please rebase to resolve conflict.

**Comment by xiangpengzhao** [comment]
Rebased. PTAL. @mengqiy
Thanks!

**Comment by mengqiy** [comment]
Thanks for cleanup.
/lgtm

**Comment by xiangpengzhao** [comment]
/test pull-kubernetes-unit

This is a flake. I also found it in another PR yesterday. Will file an issue.

**Comment by mengqiy** [comment]
@xiangpengzhao `hack/verify-golint.sh` is failing. Please fix it.

**Comment by xiangpengzhao** [comment]
@mengqiy 

Most lint errors are fixed except below two:
```
Errors from golint:
pkg/util/file/file.go:24:6: func name will be used as file.FileExists by other packages, and that stutters; consider calling this Exists
pkg/util/file/file.go:34:6: func name will be used as file.FileOrSymlinkExists by other packages, and that stutters; consider calling this OrSymlinkExists
```

I don't thinks it's so necessary to fix them. So I put `pkg/util/file` into `hack/.golint_failures` to ignore the file.

The diff in this fix is as below, PTAL. Thanks!
```
cloud@ubuntu:~/go/src/k8s.io/kubernetes$ git diff
diff --git a/hack/.golint_failures b/hack/.golint_failures
index fc6cad7..47de86c 100644
--- a/hack/.golint_failures
+++ b/hack/.golint_failures
@@ -431,7 +431,6 @@ pkg/security/podsecuritypolicy/user
 pkg/security/podsecuritypolicy/util
 pkg/securitycontext
 pkg/ssh
-pkg/util
 pkg/util/bandwidth
 pkg/util/config
 pkg/util/configz
@@ -439,6 +438,7 @@ pkg/util/dbus
 pkg/util/ebtables
 pkg/util/env
 pkg/util/exec
+pkg/util/file
 pkg/util/goroutinemap/exponentialbackoff
 pkg/util/initsystem
 pkg/util/ipconfig
diff --git a/pkg/util/file/file.go b/pkg/util/file/file.go
index bd0892d..70d26c4 100644
--- a/pkg/util/file/file.go
+++ b/pkg/util/file/file.go
@@ -20,6 +20,7 @@ import (
        "os"
 )
 
+// FileExists checks if specified file exists.
 func FileExists(filename string) (bool, error) {
        if _, err := os.Stat(filename); os.IsNotExist(err) {
                return false, nil
@@ -29,6 +30,7 @@ func FileExists(filename string) (bool, error) {
        return true, nil
 }
 
+// FileOrSymlinkExists checks if specified file or symlink exists.
 func FileOrSymlinkExists(filename string) (bool, error) {
        if _, err := os.Lstat(filename); os.IsNotExist(err) {
                return false, nil
diff --git a/pkg/util/pointer/pointer.go b/pkg/util/pointer/pointer.go
index 5b9cf0d..1a10939 100644
--- a/pkg/util/pointer/pointer.go
+++ b/pkg/util/pointer/pointer.go
@@ -21,7 +21,7 @@ import (
        "reflect"
 )
 
-// Tests whether all pointer fields in a struct are nil.  This is useful when,
+// AllPtrFieldsNil tests whether all pointer fields in a struct are nil.  This is useful when,
 // for example, an API struct is handled by plugins which need to distinguish
 // "no plugin accepted this spec" from "this spec is empty".
 //
diff --git a/pkg/util/template/template.go b/pkg/util/template/template.go
index be44ff3..2c03911 100644
--- a/pkg/util/template/template.go
+++ b/pkg/util/template/template.go
@@ -41,6 +41,7 @@ func ExecuteTemplate(w io.Writer, templateText string, data interface{}) error {
        return t.Execute(w, data)
 }
 
+// ExecuteTemplateToString executes templateText with data and output written to string.
 func ExecuteTemplateToString(templateText string, data interface{}) (string, error) {
        b := bytes.Buffer{}
        err := ExecuteTemplate(&b, templateText, data)
(END)
```

**Comment by mengqiy** [comment]
/lgtm
Thanks for giving me the diff.

**Pull request merged**